lib/puppet/resource.rb

```diff
@@ -627,10 +627,10 @@ def self.type_and_title(type, title)
 
   def self.extract_type_and_title(argtype, argtitle)
     if (argtype.nil? || argtype == :component || argtype == :whit) &&
-          argtitle =~ /^([^\[\]]+)\[(.+)\]$/m                  then [ $1,                 $2            ]
-    elsif argtitle.nil? && argtype =~ /^([^\[\]]+)\[(.+)\]$/m  then [ $1,                 $2            ]
-    elsif argtitle                                             then [ argtype,            argtitle      ]
-    elsif argtype.is_a?(Puppet::Type)                          then [ argtype.class.name, argtype.title ]
+          argtitle =~ /^([^\[\]]+)\[(.+)\]$/m                       then [ $1,                 $2            ]
+    elsif argtitle.nil? && argtype.to_s =~ /^([^\[\]]+)\[(.+)\]$/m  then [ $1,                 $2            ]
```

Comment:
I think this case is trying to match calls like `Resource.new("File[/path/to/thing]")` where `argtitle` is nil and `argtype` is a String. If we call `to_s` then we could accidentally match something that isn't a string, but whose `to_s` method happens to produce something that matches the regex. Maybe better to check for that case explicitly like `elsif argtitle.nil? && argtype.is_a?(String) && argtype =~ /..../m`?